File: Go/sort/merge_sort.go
package sort

func MergeSort(arr []int) {
	Divide(arr, 0, len(arr)-1)
}

func Divide(arr []int, left int, right int) {
	if left < right {
		mid := (left + right) / 2
		Divide(arr, left, mid)
		Divide(arr, mid+1, right)
		Conquer(arr, left, right)
	}
}

func Conquer(arr []int, left int, right int) {
	sorted := make([]int, right-left+1)
	low := left
	mid := (left + right) / 2
	high := mid + 1
	index := 0

	for low <= mid && high <= right {
		if arr[low] < arr[high] {
			sorted[index] = arr[low]
			index++
			low++
		} else {
			sorted[index] = arr[high]
			index++
			high++
		}
	}
	for low <= mid {
		sorted[index] = arr[low]
		index++
		low++
	}
	for high <= right {
		sorted[index] = arr[high]
		index++
		high++
	}
	for i := left; i <= right; i++ {
		arr[i] = sorted[i-left]
	}
}

File: Python/sort/MergeSort.py
class MergeSort:
    def __init__(self, arr):
        self.arr = arr
        self.length = len(arr)
        self.sort(0, self.length - 1)

    def sort(self, left, right):
        if left < right:
            mid = (left + right) // 2
            self.sort(left, mid)
            self.sort(mid + 1, right)
            self.merge(left, right)

    def merge(self, left, right):
        mid = (left + right) // 2
        low, high, index = left, mid + 1, 0
        sorted_arr = [0] * (right - left + 1)
        while low <= mid and high <= right:
            if self.arr[low] < self.arr[high]:
                sorted_arr[index] = self.arr[low]
                low += 1
            else:
                sorted_arr[index] = self.arr[high]
                high += 1
            index += 1

        for i in range(low, mid + 1):
            sorted_arr[index] = self.arr[i]
            index += 1

        for i in range(high, right + 1):
            sorted_arr[index] = self.arr[i]
            index += 1

        for i in range(left, right + 1):
            self.arr[i] = sorted_arr[i - left]
